lib/evmone/eof.cpp

```diff
@@ -13,6 +13,92 @@ namespace
 {
 constexpr uint8_t FORMAT = 0xef;
 constexpr uint8_t MAGIC = 0x00;
+constexpr uint8_t TERMINATOR = 0x00;
+constexpr uint8_t CODE_SECTION = 0x01;
+constexpr uint8_t DATA_SECTION = 0x02;
+
+std::pair<EOF1Header, EOFValidationErrror> validate_eof_headers(bytes_view code) noexcept
+{
+    enum class State
+    {
+        section_id,
+        section_size,
+        terminated
+    };
+
+    auto state = State::section_id;
+    uint8_t section_id = 0;
+    size_t section_sizes[3] = {0, 0, 0};
+    const auto code_end = code.end();
+    auto it = code.begin() + sizeof(MAGIC) + 2;  // FORMAT + MAGIC + VERSION
+    while (it != code_end && state != State::terminated)
+    {
+        switch (state)
+        {
+        case State::section_id:
+        {
+            section_id = *it;
+            switch (section_id)
+            {
+            case TERMINATOR:
+                if (section_sizes[CODE_SECTION] == 0)
+                    return {{}, EOFValidationErrror::code_section_missing};
+                state = State::terminated;
+                break;
+            case DATA_SECTION:
+                if (section_sizes[CODE_SECTION] == 0)
+                    return {{}, EOFValidationErrror::code_section_missing};
+                if (section_sizes[DATA_SECTION] != 0)
+                    return {{}, EOFValidationErrror::multiple_data_sections};
+                state = State::section_size;
+                break;
+            case CODE_SECTION:
+                if (section_sizes[CODE_SECTION] != 0)
+                    return {{}, EOFValidationErrror::multiple_code_sections};
+                state = State::section_size;
+                break;
+            default:
+                return {{}, EOFValidationErrror::unknown_section_id};
+            }
+            break;
+        }
+        case State::section_size:
+        {
+            const auto size_hi = *it;
+            ++it;
+            if (it == code_end)
+                return {{}, EOFValidationErrror::incomplete_section_size};
+            const auto size_lo = *it;
+            const auto section_size = static_cast<size_t>(size_hi << 8) | size_lo;
+            if (section_size == 0)
+                return {{}, EOFValidationErrror::zero_section_size};
+
+            section_sizes[section_id] = section_size;
+            state = State::section_id;
+            break;
+        }
+        case State::terminated:
+            return {{}, EOFValidationErrror::impossible};
+        }
+
+        ++it;
```

Comment:
I think this may be better moved to the cases, ie. `size_hi = *it++` ?

Reply:
Moved.